### ISSUE TITLE
Final Project Feedback

### ISSUE THREAD

**Issue by ConsenSys-Academy** [tonybutt/blockchain-developer-bootcamp-final-project#1]
Hey @tonybutt – Congratulations, your Final Project passed! We'll be approving it on the course itself in a week or two, but here's your feedback in the meanwhile:

### Frontend

- Good to have either a MetaMask "Switch Network" button or an easier way to users to switch or at least not render content before
- I did have some issues with the page reloading – had to refresh to see the pipeline populated – you think it's a routing issue with Netlify? Good to check and have running -- also might be good to have a loading animation once you detect a txn has been submitted, that way the user is getting some visual feedback that there's work being done on the backend.
- Frontend looks really solid and appreciate you not being able to click to tabs like ""Owner Area""
- Great job getting all this off the ground

### Smart Contract Design and Security

- Nice separation of contract and client in the repository, helps with navigation of the project
- Ditto on the commenting in the contracts, really helps with comprehension of someone reading the code
- You mention ERC-20 Pausable in your Design Patterns, but I didn't see it imported or used in `SecureCode.sol` – is this an artifact of an earlier version? I'm seeing that you use Pausable access control for the ERC-721
- I like how tight the code is for functions like `approveCode` and others. That in and of itself can be a really nice security feature, decreasing surface area for attacks / bugs
- Good you took the steps of explicitly stating visibility for functions and locking your pragma version

### Testing

- Tests look solid. Again, the concise nature of the programming is excellent. Being able to boil down to the essential logic is so critical and I think can be very challenging. It also makes reading the tests very straightforward.
- Also like the definition and use of `PIPELINE_STUB`.

### Overall

- I'm always really interested in the incorporation of APIs from sites like Gitlab with public blockchains. This is a cool add-on which doesn't disturb (or even inform) Gitlab.
- Really interesting concept about giving folks an NFT for contributing to a project – This is for folks who have a successful PR, is that right? And the owner who merges it to the Gitlab repo can also approve their deployment on-chain as well? Curious how you came up with this. It's a really cool idea, @tonybutt!

**Comment by tonybutt** [comment]
Yeah I am tracking the overall UX issues, and honestly the whole client not being tested is a bit sub-par for me as well. Not gonna lie that I did nearly bare minimum to pass and some of the outdated Docs reflect my lack of time to refine. 

Overall I'm glad you think it's a cool idea. I'm a Software Engineer with a huge specialty in DevSecOps and modern deployment platforms. Having the traceability of commits within the system itself is huge. Blockchain brings this visibility. Ideally, each process of a deployment pipeline would sign the commit hash. This helps with supply chain attacks as knowing where the code is coming from(especially depends) and who is building/compiling the code is imperative to having a secure environment. Blockchain's tamper evidence helps immensely here to keep nefarious actors out of our critical systems.

**Comment by tonybutt** [comment]
And by automating the process we achieve a higher posture for security without compromising the development experience which is really hard when you ponder about it.
